pkg/controllers/provisioning/v1alpha1/reallocation/terminate.go

```diff
@@ -19,18 +19,70 @@ import (
 	"fmt"
 	"github.com/awslabs/karpenter/pkg/apis/provisioning/v1alpha1"
 	"github.com/awslabs/karpenter/pkg/cloudprovider"
+	"github.com/awslabs/karpenter/pkg/utils/ptr"
+	"github.com/awslabs/karpenter/pkg/utils/scheduling"
 	"go.uber.org/zap"
 	v1 "k8s.io/api/core/v1"
+	"k8s.io/api/policy/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
 type Terminator struct {
 	kubeClient    client.Client
 	cloudprovider cloudprovider.Factory
+	coreV1Client  corev1.CoreV1Interface
+}
+
+// DrainNodes takes in a list of expired non-drained nodes and calls drain on them
+func (t *Terminator) DrainNodes(ctx context.Context, nodes []*v1.Node, provisioner *v1alpha1.Provisioner) error {
+	for _, node := range nodes {
+		// (0.) TODO: Check if Node should be drained
+		// - Disrupts PDB
+		// - Pods owned by controller object
+		// - Pod on Node can't be rescheduled elsewhere
+
+		// 1. Get Pods on Node
+		podList := &v1.PodList{}
+		if err := t.kubeClient.List(ctx, podList, client.MatchingFields{"spec.nodeName": node.Name}); err != nil {
+			return fmt.Errorf("listing pods for node %s, %w", node.Name, err)
+		}
+		// 1a. If no pods on node, label as drained
+		empty := true
+		for _, pod := range ptr.PodListToSlice(podList) {
+			if !scheduling.IsOwnedByDaemonSet(pod) {
+				empty = false
+				if err := t.coreV1Client.Pods(pod.Namespace).Evict(ctx, &v1beta1.Eviction{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: pod.Name,
+					},
+				}); err != nil {
+					return fmt.Errorf("evicting pods from node %s, %w", node.Name, err)
+				}
+			}
+		}
+		persisted := node.DeepCopy()
+		if !empty {
+			// Mark as drained
+			node.Labels[v1alpha1.ProvisionerDrainingKey] = "draining"
+		} else {
+			// Mark Node as Draining
+			node.Labels[v1alpha1.ProvisionerDrainingKey] = "drained"
```

Comment:
Why not just delete the node immediately, rather than mark as drained and then delete?